src/clickhouseEngine.ts

```diff
@@ -0,0 +1,231 @@
+import { asDate, asObject, asOptional, asString, uncleaner } from 'cleaners'
+import nano from 'nano'
+
+import { config } from './config'
+import { initDbs } from './initDbs'
+import { asStandardTx, DbTx, StandardTx, wasDbTx } from './types'
+import { datelog, snooze } from './util'
+
+// Clickhouse recommends large batch inserts. We consider the couchdb
+// query size as well.
+const PAGE_SIZE = 10_000
+
+const nanoDb = nano(config.couchDbFullpath)
+const clickhouseDb = createClient(config.clickhouseConnection)
+
+const progressDocName = 'clickhouse:clickhouseEngine'
+const asClickhouseProgress = asObject({
+  _id: asOptional(asString, progressDocName),
+  _rev: asOptional(asString),
+  afterTime: asOptional(asDate, new Date(0))
+})
+const wasClickhouseProgress = uncleaner(asClickhouseProgress)
+
+export async function clickhouseEngine(): Promise<void> {
+  await initDbs()
+  await initClickhouseDb()
+
+  const dbTransactions = nanoDb.db.use<StandardTx>('reports_transactions')
+
+  const dbProgress = nanoDb.db.use('reports_progresscache')
+  const out = await dbProgress.get(progressDocName).catch(error => {
+    if (error.error != null && error.error === 'not_found') {
+      datelog(`Previous Progress Record Not Found ${progressDocName}`)
+      return {}
+    }
+    throw error
+  })
+  const progressDoc = asClickhouseProgress(out)
+
+  let bookmark: string | undefined
+
+  while (true) {
+    const response = await dbTransactions.find({
+      selector: {
+        status: { $eq: 'complete' },
+        updateTime: { $gt: progressDoc.afterTime.toISOString() }
+      },
+      sort: [{ updateTime: 'asc' }],
+      use_index: 'status-updatetime',
+      limit: PAGE_SIZE,
+      bookmark
+    })
+
+    bookmark = response.bookmark
+
+    const startDocId = response.docs[0]?._id
+    const endDocId = response.docs[response.docs.length - 1]?._id
+
+    if (response.docs.length > 0) {
+      datelog(
+        `Processing ${response.docs.length} rows from ${startDocId} to ${endDocId}.`
+      )
+    } else {
+      datelog(
+        `Queried for new documents after ${progressDoc.afterTime.toISOString()}.`
+      )
+    }
+
+    const newDocs: DbTx[] = []
+    const newRows: any[][] = []
+    let lastDocUpdateTime: string | undefined
+
+    for (const doc of response.docs) {
+      const { appId, partnerId } = getDocumentIdentifiers(doc._id)
+
+      const standardTx = asStandardTx(doc)
+
+      newRows.push([
+        appId,
+        partnerId,
+        standardTx.orderId,
+        standardTx.countryCode,
+        standardTx.depositTxid,
+        standardTx.depositAddress,
+        standardTx.depositCurrency,
+        standardTx.depositAmount,
+        standardTx.direction,
+        standardTx.exchangeType,
+        standardTx.paymentType,
+        standardTx.payoutTxid,
+        standardTx.payoutAddress,
+        standardTx.payoutCurrency,
+        standardTx.payoutAmount,
+        standardTx.status,
+        Math.round(standardTx.timestamp),
```

Comment:
why round. our timestamp is in seconds so there should be a few decimals to account for MS.